File: tiktok/apis/data_portability.py
"""Data Portability API namespace.

Wraps the following TikTok endpoints:

* ``POST /v2/user/data/add/``      — initiate a data export request
* ``POST /v2/user/data/check/``    — check the status of a request
* ``POST /v2/user/data/cancel/``   — cancel a pending request
* ``POST /v2/user/data/download/`` — download the exported zip archive

Reference: https://developers.tiktok.com/doc/data-portability-api-get-started

Availability note: This API is available only to users in the **European
Economic Area (EEA)** or the **United Kingdom (UK)**.
"""

from __future__ import annotations

from typing import Any

from tiktok.apis.base import BaseAPI
from tiktok.config import TIKTOK_API_BASE_URL
from tiktok.models.data_portability import (
    AddDataRequestData,
    DataCategory,
    DataFormat,
    DataRequestStatusData,
    StatusField,
)

_BASE = TIKTOK_API_BASE_URL

# Fields to request from the status endpoint by default.
_ALL_STATUS_FIELDS = list(StatusField)


class DataPortabilityAPI(BaseAPI):
    """Methods for exporting and downloading a user's TikTok data.

    All methods require an access token authorised with one or more
    ``portability.*`` scopes corresponding to the requested data categories.
    """

    # ------------------------------------------------------------------
    # Add data request
    # ------------------------------------------------------------------

    async def add_data_request(
        self,
        *,
        data_format: DataFormat,
        category_selection_list: list[DataCategory],
    ) -> AddDataRequestData:
        """Initiate a new data export request.

        Parameters
        ----------
        data_format:
            Output format for the archive (``"json"`` or ``"text"``).
        category_selection_list:
            One or more data categories to include in the export.  The
            authenticated user must have granted the corresponding
            ``portability.*`` scope(s).

        Scope required: varies by category (see :class:`DataCategory`).

        Returns:
            :class:`~tiktok.models.data_portability.AddDataRequestData`
            containing the ``request_id`` needed for subsequent calls.

        Raises:
            :class:`~tiktok.exceptions.TikTokAuthError`: Token lacks scope.
            :class:`~tiktok.exceptions.TikTokAPIError`: Any other API error.
        """
        payload = await self._session.post(
            f"{_BASE}/v2/user/data/add/",
            json={
                "data_format": data_format.value,
                "category_selection_list": [c.value for c in category_selection_list],
            },
            params={"fields": "request_id"},
        )
        return AddDataRequestData.model_validate(payload["data"])

    # ------------------------------------------------------------------
    # Check status
    # ------------------------------------------------------------------

    async def check_data_request_status(
        self,
        request_id: int,
        *,
        fields: list[StatusField] | None = None,
    ) -> DataRequestStatusData:
        """Check the processing status of an export request.

        Parameters
        ----------
        request_id:
            The ``request_id`` returned by :meth:`add_data_request`.
        fields:
            Status fields to include in the response.  Defaults to all
            available fields (see :class:`StatusField`).

        Scope required: same as used when creating the request.

        Returns:
            :class:`~tiktok.models.data_portability.DataRequestStatusData`.

        Raises:
            :class:`~tiktok.exceptions.TikTokAuthError`: Token lacks scope.
            :class:`~tiktok.exceptions.TikTokAPIError`: Any other API error.
        """
        requested_fields: list[StatusField] = fields if fields is not None else _ALL_STATUS_FIELDS
        params: dict[str, Any] = {
            "fields": ",".join(f.value for f in requested_fields),
        }
        payload = await self._session.post(
            f"{_BASE}/v2/user/data/check/",
            json={"request_id": request_id},
            params=params,
        )
        return DataRequestStatusData.model_validate(payload["data"])

    # ------------------------------------------------------------------
    # Cancel data request
    # ------------------------------------------------------------------

    async def cancel_data_request(self, request_id: int) -> None:
        """Cancel a pending data export request.

        Parameters
        ----------
        request_id:
            The ``request_id`` to cancel.

        Scope required: same as used when creating the request.

        Raises:
            :class:`~tiktok.exceptions.TikTokAuthError`: Token lacks scope.
            :class:`~tiktok.exceptions.TikTokAPIError`: Any other API error.
        """
        await self._session.post(
            f"{_BASE}/v2/user/data/cancel/",
            params={"request_id": str(request_id)},
        )

    # ------------------------------------------------------------------
    # Download data
    # ------------------------------------------------------------------

    async def download_data(self, request_id: int) -> bytes:
        """Download the exported data archive as raw bytes.

        The archive is a ``.zip`` file.  Call this only after the request
        status is ``DataRequestStatus.DOWNLOADING`` (or you have received
        the webhook notification from TikTok).

        Parameters
        ----------
        request_id:
            The ``request_id`` returned by :meth:`add_data_request`.

        Scope required: same as used when creating the request.

        Returns:
            Raw bytes of the zip archive.

        Raises:
            :class:`~tiktok.exceptions.TikTokAuthError`: Token lacks scope.
            :class:`~tiktok.exceptions.TikTokUploadError`: Download failure.
            :class:`~tiktok.exceptions.TikTokAPIError`: Any other API error.
        """
        return await self._session.post_stream(
            f"{_BASE}/v2/user/data/download/",
            json={"request_id": request_id},
        )

File: go/data_portability.go
package tiktok

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// Data Portability API.
//
// These endpoints let an authenticated user export and download their TikTok
// data. They wrap:
//
//   - POST /v2/user/data/add/      — initiate a data export request
//   - POST /v2/user/data/check/    — check the status of a request
//   - POST /v2/user/data/cancel/   — cancel a pending request
//   - POST /v2/user/data/download/ — download the exported zip archive
//
// Availability note: the Data Portability API is available only to users in the
// European Economic Area (EEA) or the United Kingdom (UK). Calls from other
// regions are rejected upstream regardless of token validity.
//
// Reference: https://developers.tiktok.com/doc/data-portability-api-get-started

// DataCategory enumerates the data categories that can be requested via the
// Data Portability API. Each category corresponds to one or more portability.*
// OAuth scopes that the user must have granted.
type DataCategory string

const (
	// DataCategoryAllData maps to portability.all.single / portability.all.ongoing.
	DataCategoryAllData DataCategory = "all_data"
	// DataCategoryActivity maps to portability.activity.single / .ongoing.
	DataCategoryActivity DataCategory = "activity"
	// DataCategoryVideo maps to portability.postsandprofile.single / .ongoing.
	DataCategoryVideo DataCategory = "video"
	// DataCategoryProfile maps to portability.postsandprofile.single / .ongoing.
	DataCategoryProfile DataCategory = "profile"
	// DataCategoryDirectMessage maps to portability.directmessages.single / .ongoing.
	DataCategoryDirectMessage DataCategory = "direct_message"
)

// DataFormat enumerates the output format for a data export archive.
type DataFormat string

const (
	// DataFormatText requests the export in plain-text format.
	DataFormatText DataFormat = "text"
	// DataFormatJSON requests the export in JSON format.
	DataFormatJSON DataFormat = "json"
)

// DataRequestStatus enumerates the processing states of a data export request.
type DataRequestStatus string

const (
	// DataRequestStatusPending indicates the export is still being prepared.
	DataRequestStatusPending DataRequestStatus = "pending"
	// DataRequestStatusDownloading indicates the archive is ready to download.
	DataRequestStatusDownloading DataRequestStatus = "downloading"
	// DataRequestStatusExpired indicates the export window has elapsed.
	DataRequestStatusExpired DataRequestStatus = "expired"
	// DataRequestStatusCancelled indicates the request was cancelled.
	DataRequestStatusCancelled DataRequestStatus = "cancelled"
)

// StatusField enumerates the fields that can be requested from
// /v2/user/data/check/ via the fields query parameter.
type StatusField string

const (
	// StatusFieldRequestID requests the request_id field.
	StatusFieldRequestID StatusField = "request_id"
	// StatusFieldApplyTime requests the apply_time field.
	StatusFieldApplyTime StatusField = "apply_time"
	// StatusFieldCollectTime requests the collect_time field.
	StatusFieldCollectTime StatusField = "collect_time"
	// StatusFieldStatus requests the status field.
	StatusFieldStatus StatusField = "status"
	// StatusFieldDataFormat requests the data_format field.
	StatusFieldDataFormat StatusField = "data_format"
	// StatusFieldCategorySelectionList requests the category_selection_list field.
	StatusFieldCategorySelectionList StatusField = "category_selection_list"
)

// allStatusFields is the default set requested by CheckDataRequestStatus when no
// explicit field list is supplied. The order mirrors the Python SDK so the
// serialised fields query parameter is byte-identical.
var allStatusFields = []StatusField{
	StatusFieldRequestID,
	StatusFieldApplyTime,
	StatusFieldCollectTime,
	StatusFieldStatus,
	StatusFieldDataFormat,
	StatusFieldCategorySelectionList,
}

// AddDataRequestResult is the response data from /v2/user/data/add/.
type AddDataRequestResult struct {
	// RequestID is the unique identifier for the export request. Persist it; it
	// is required by the status-check, cancel, and download endpoints.
	RequestID int64 `json:"request_id"`
}

// DataRequestStatusResult is the response data from /v2/user/data/check/.
//
// Only fields included in the fields query parameter are populated by TikTok;
// all others are left at their zero value. Pointer fields distinguish "field
// not requested / not returned" (nil) from a returned zero value.
type DataRequestStatusResult struct {
	RequestID             *int64             `json:"request_id,omitempty"`
	ApplyTime             *int64             `json:"apply_time,omitempty"`
	CollectTime           *int64             `json:"collect_time,omitempty"`
	Status                *DataRequestStatus `json:"status,omitempty"`
	DataFormat            *DataFormat        `json:"data_format,omitempty"`
	CategorySelectionList []DataCategory     `json:"category_selection_list,omitempty"`
}

// AddDataRequestParams configures AddDataRequest.
type AddDataRequestParams struct {
	// DataFormat is the output format for the archive ("json" or "text").
	DataFormat DataFormat
	// CategorySelectionList is one or more data categories to include in the
	// export. The user must have granted the corresponding portability.* scopes.
	CategorySelectionList []DataCategory
}

// AddDataRequest initiates a new data export request.
//
// It POSTs to /v2/user/data/add/ with the requested format and categories, and
// always passes fields=request_id as a query parameter (mirroring the Python
// SDK). The returned RequestID is required by all subsequent calls.
//
// Availability: EEA / UK users only.
//
// Scope: varies by category (see DataCategory).
func (c *Client) AddDataRequest(ctx context.Context, p AddDataRequestParams) (*AddDataRequestResult, error) {
	if len(p.CategorySelectionList) == 0 {
		return nil, errors.New("tiktok: AddDataRequest requires at least one category")
	}
	categories := make([]string, len(p.CategorySelectionList))
	for i, cat := range p.CategorySelectionList {
		categories[i] = string(cat)
	}
	body := map[string]any{
		"data_format":             string(p.DataFormat),
		"category_selection_list": categories,
	}
	q := url.Values{}
	q.Set("fields", "request_id")

	data, err := c.doJSON(ctx, http.MethodPost, c.baseURL+"/v2/user/data/add/", body, q)
	if err != nil {
		return nil, err
	}
	out := &AddDataRequestResult{}
	if err := json.Unmarshal(data, out); err != nil {
		return nil, fmt.Errorf("tiktok: decode add data request: %w", err)
	}
	return out, nil
}

// CheckDataRequestStatus checks the processing status of an export request.
//
// It POSTs to /v2/user/data/check/ with the request_id in the JSON body and the
// requested status fields as a comma-separated fields query parameter. When
// fields is nil, all available status fields are requested (matching the Python
// SDK default).
//
// Availability: EEA / UK users only.
//
// Scope: same scopes used when creating the request.
func (c *Client) CheckDataRequestStatus(ctx context.Context, requestID int64, fields []StatusField) (*DataRequestStatusResult, error) {
	requested := fields
	if requested == nil {
		requested = allStatusFields
	}
	fieldStrings := make([]string, len(requested))
	for i, f := range requested {
		fieldStrings[i] = string(f)
	}
	q := url.Values{}
	q.Set("fields", strings.Join(fieldStrings, ","))

	body := map[string]any{"request_id": requestID}
	data, err := c.doJSON(ctx, http.MethodPost, c.baseURL+"/v2/user/data/check/", body, q)
	if err != nil {
		return nil, err
	}
	out := &DataRequestStatusResult{}
	if err := json.Unmarshal(data, out); err != nil {
		return nil, fmt.Errorf("tiktok: decode data request status: %w", err)
	}
	return out, nil
}

// CancelDataRequest cancels a pending data export request.
//
// It POSTs to /v2/user/data/cancel/ with the request_id as a query parameter
// (no JSON body), mirroring the Python SDK. It returns nil on success.
//
// Availability: EEA / UK users only.
//
// Scope: same scopes used when creating the request.
func (c *Client) CancelDataRequest(ctx context.Context, requestID int64) error {
	q := url.Values{}
	q.Set("request_id", strconv.FormatInt(requestID, 10))

	_, err := c.doJSON(ctx, http.MethodPost, c.baseURL+"/v2/user/data/cancel/", nil, q)
	return err
}

// DownloadData downloads the exported data archive as raw bytes.
//
// It POSTs to /v2/user/data/download/ with the request_id in the JSON body and
// returns the complete zip archive. Call it only once the request status is
// DataRequestStatusDownloading (or after the TikTok webhook notification).
// Errors are still surfaced as *tiktok.Error from the JSON error envelope.
//
// Availability: EEA / UK users only.
//
// Scope: same scopes used when creating the request.
func (c *Client) DownloadData(ctx context.Context, requestID int64) ([]byte, error) {
	body := map[string]any{"request_id": requestID}
	return c.doRaw(ctx, http.MethodPost, c.baseURL+"/v2/user/data/download/", body, nil)
}
